File: too_many_cooks/lib/src/admin_routes.dart
/// Admin REST endpoints for the VSCode extension.
///
/// The VSIX talks to these endpoints — never touches the DB directly.
/// Streamable HTTP endpoint pushes all state changes in real-time.
library;

import 'dart:async';
import 'dart:js_interop';
import 'dart:js_interop_unsafe';

import 'package:dart_node_core/dart_node_core.dart';
import 'package:dart_node_express/dart_node_express.dart';
import 'package:dart_node_mcp/dart_node_mcp.dart';
import 'package:nadz/nadz.dart';
import 'package:too_many_cooks_data/too_many_cooks_data.dart';

/// Admin event hub — manages Streamable HTTP transports
/// for pushing real-time events to the VSIX.
typedef AdminEventHub = ({
  Map<String, StreamableHttpTransport> transports,
  Map<String, McpServer> servers,
  void Function(String event, Map<String, Object?> payload)
      pushEvent,
});

/// Create an admin event hub for Streamable HTTP push.
AdminEventHub createAdminEventHub() {
  final transports =
      <String, StreamableHttpTransport>{};
  final servers = <String, McpServer>{};

  void pushEvent(
    String event,
    Map<String, Object?> payload,
  ) {
    consoleError(
      '[TMC] [PUSH] $event → '
      '${servers.length} server(s), '
      '${transports.length} transport(s)',
    );
    final data = <String, Object?>{
      'event': event,
      'timestamp': DateTime.now().millisecondsSinceEpoch,
      'payload': payload,
    };

    for (final entry in [...servers.entries]) {
      consoleError(
        '[TMC] [PUSH] Sending to ${entry.key}',
      );
      unawaited(
        entry.value
            .sendLoggingMessage((
              level: 'info',
              logger: 'too-many-cooks-admin',
              data: data,
            ))
            .then((_) {
              consoleError(
                '[TMC] [PUSH] Sent OK to ${entry.key}',
              );
            }, onError: (Object e) {
              consoleError(
                '[TMC] [PUSH] FAILED ${entry.key}: $e',
              );
              // Transport closed — remove it
              servers.remove(entry.key);
              transports.remove(entry.key);
            }),
      );
    }
  }

  return (
    transports: transports,
    servers: servers,
    pushEvent: pushEvent,
  );
}

/// Register admin routes on an Express app.
void registerAdminRoutes(
  ExpressApp app,
  TooManyCooksDb db,
  AdminEventHub hub,
) {
  // JSON body parser
  final expressModule = requireModule('express') as JSObject;
  final jsonMiddleware =
      (expressModule['json'] as JSFunction?)
          ?.callAsFunction(expressModule);
  app
    ..use(jsonMiddleware)

    // GET /admin/status — full status snapshot
    ..get('/admin/status', handler((req, res) {
      final agents = switch (db.listAgents()) {
        Success(:final value) =>
          value.map(agentIdentityToJson).join(','),
        Error() => '',
      };
      final locks = switch (db.listLocks()) {
        Success(:final value) =>
          value.map(fileLockToJson).join(','),
        Error() => '',
      };
      final plans = switch (db.listPlans()) {
        Success(:final value) =>
          value.map(agentPlanToJson).join(','),
        Error() => '',
      };
      final messages = switch (db.listAllMessages()) {
        Success(:final value) =>
          value.map(messageToJson).join(','),
        Error() => '',
      };

      res
        ..set('Content-Type', 'application/json')
        ..send(
          '{"agents":[$agents],"locks":[$locks],'
          '"plans":[$plans],"messages":[$messages]}',
        );
    }))

    // POST /admin/delete-lock — force-delete a lock
    ..post('/admin/delete-lock', handler((req, res) {
      final body = _parseBody(req);
      final filePath = body['filePath'] as String?;
      if (filePath == null) {
        _sendError(res, 400, 'filePath required');
        return;
      }
      switch (db.adminDeleteLock(filePath)) {
        case Success():
          hub.pushEvent(
            'lock_released',
            {'file_path': filePath},
          );
          res.send('{"deleted":true}');
        case Error(:final error):
          _sendError(res, 400, dbErrorToJson(error));
      }
    }))

    // POST /admin/delete-agent — delete agent + data
    ..post('/admin/delete-agent', handler((req, res) {
      final body = _parseBody(req);
      final agentName = body['agentName'] as String?;
      if (agentName == null) {
        _sendError(res, 400, 'agentName required');
        return;
      }
      switch (db.adminDeleteAgent(agentName)) {
        case Success():
          hub.pushEvent(
            'agent_deleted',
            {'agent_name': agentName},
          );
          res.send('{"deleted":true}');
        case Error(:final error):
          _sendError(res, 400, dbErrorToJson(error));
      }
    }))

    // POST /admin/reset-key — generate new key for agent
    ..post('/admin/reset-key', handler((req, res) {
      final body = _parseBody(req);
      final agentName = body['agentName'] as String?;
      if (agentName == null) {
        _sendError(res, 400, 'agentName required');
        return;
      }
      switch (db.adminResetKey(agentName)) {
        case Success(:final value):
          res.send(agentRegistrationToJson(value));
        case Error(:final error):
          _sendError(res, 400, dbErrorToJson(error));
      }
    }))

    // POST /admin/send-message — send message (no auth)
    ..post('/admin/send-message', handler((req, res) {
      final body = _parseBody(req);
      final fromAgent = body['fromAgent'] as String?;
      final toAgent = body['toAgent'] as String?;
      final content = body['content'] as String?;
      if (fromAgent == null ||
          toAgent == null ||
          content == null) {
        _sendError(
          res,
          400,
          'fromAgent, toAgent, content required',
        );
        return;
      }
      switch (db.adminSendMessage(
        fromAgent,
        toAgent,
        content,
      )) {
        case Success(:final value):
          hub.pushEvent('message_sent', {
            'from_agent': fromAgent,
            'to_agent': toAgent,
            'message_id': value,
          });
          res.send(
            '{"sent":true,"message_id":"$value"}',
          );
        case Error(:final error):
          _sendError(res, 400, dbErrorToJson(error));
      }
    }))

    // POST /admin/reset — clear all data (testing)
    ..post('/admin/reset', handler((req, res) {
      switch (db.adminReset()) {
        case Success():
          hub.pushEvent('state_reset', <String, Object?>{});
          res.send('{"reset":true}');
        case Error(:final error):
          _sendError(res, 500, dbErrorToJson(error));
      }
    }));
}

/// Send an error response.
void _sendError(Response res, int code, String message) {
  res
    ..status(code)
    ..send(message);
}

/// Parse request body as Map.
Map<String, Object?> _parseBody(Request req) {
  final body = req.body;
  if (body == null) return {};
  final dartified = body.dartify();
  if (dartified is Map) {
    return Map<String, Object?>.fromEntries(
      dartified.entries.map(
        (e) => MapEntry(e.key.toString(), e.value),
      ),
    );
  }
  return {};
}

File: too-many-cooks/lib/src/admin_routes.ts
/// Admin REST endpoints for the VSCode extension.
///
/// The VSIX talks to these endpoints - never touches the DB directly.
/// Streamable HTTP endpoint pushes all state changes in real-time.

import express, { type Request, type Response, type Express } from "express";
import type { McpServer } from "@modelcontextprotocol/sdk/server/mcp.js";
import type { StreamableHTTPServerTransport } from "@modelcontextprotocol/sdk/server/streamableHttp.js";
import {
  type TooManyCooksDb,
  agentIdentityToJson,
  fileLockToJson,
  agentPlanToJson,
  messageToJson,
  agentRegistrationToJson,
  dbErrorToJson,
} from "./data/data.js";

/** Logger name for admin push events. */
const ADMIN_LOGGER_NAME = "too-many-cooks-admin";

/** Admin event hub - manages Streamable HTTP transports for pushing real-time events to the VSIX. */
export type AdminEventHub = {
  readonly transports: Map<string, StreamableHTTPServerTransport>;
  readonly servers: Map<string, McpServer>;
  readonly pushEvent: (
    event: string,
    payload: Record<string, unknown>,
  ) => void;
};

/** Send a push event to all connected admin servers. */
const sendToServer = async (
  entry: [string, McpServer],
  data: Record<string, unknown>,
  servers: Map<string, McpServer>,
  transports: Map<string, StreamableHTTPServerTransport>,
): Promise<void> => {
  const [key, server] = entry;
  console.error(`[TMC] [PUSH] Sending to ${key}`);
  try {
    await server.sendLoggingMessage({
      level: "info",
      logger: ADMIN_LOGGER_NAME,
      data,
    });
    console.error(`[TMC] [PUSH] Sent OK to ${key}`);
  } catch {
    console.error(`[TMC] [PUSH] FAILED ${key}`);
    servers.delete(key);
    transports.delete(key);
  }
};

/** Create an admin event hub for Streamable HTTP push. */
export const createAdminEventHub = (): AdminEventHub => {
  const transports = new Map<string, StreamableHTTPServerTransport>();
  const servers = new Map<string, McpServer>();

  const pushEvent = (
    event: string,
    payload: Record<string, unknown>,
  ): void => {
    console.error(
      `[TMC] [PUSH] ${event} -> ${String(servers.size)} server(s), ${String(transports.size)} transport(s)`,
    );
    const data: Record<string, unknown> = {
      event,
      timestamp: Date.now(),
      payload,
    };
    // eslint-disable-next-line @typescript-eslint/no-floating-promises
    Promise.all(
      [...servers.entries()].map(
        // eslint-disable-next-line require-await
        async (entry) => sendToServer(entry, data, servers, transports),
      ),
    );
  };

  return { transports, servers, pushEvent };
};

/** Send an error response. */
const sendError = (res: Response, code: number, message: string): void => {
  res.status(code).send(message);
};

/** Extract a string field from the request body, or undefined. */
const stringField = (
  body: Record<string, unknown>,
  field: string,
): string | undefined => {
  const value = body[field];
  return typeof value === "string" ? value : undefined;
};

/** Register GET /admin/status route. */
const registerStatusRoute = (app: Express, db: TooManyCooksDb): void => {
  app.get("/admin/status", (_req: Request, res: Response) => {
    const agentsResult = db.listAgents();
    const agents = agentsResult.ok
      ? agentsResult.value.map(agentIdentityToJson)
      : [];
    const locksResult = db.listLocks();
    const locks = locksResult.ok
      ? locksResult.value.map(fileLockToJson)
      : [];
    const plansResult = db.listPlans();
    const plans = plansResult.ok
      ? plansResult.value.map(agentPlanToJson)
      : [];
    const messagesResult = db.listAllMessages();
    const messages = messagesResult.ok
      ? messagesResult.value.map(messageToJson)
      : [];
    res
      .set("Content-Type", "application/json")
      .send(JSON.stringify({ agents, locks, plans, messages }));
  });
};

/** Register POST /admin/delete-lock route. */
const registerDeleteLockRoute = (
  app: Express,
  db: TooManyCooksDb,
  hub: AdminEventHub,
): void => {
  app.post("/admin/delete-lock", (req: Request, res: Response) => {
    const filePath = stringField(req.body as Record<string, unknown>, "filePath");
    if (filePath === undefined) {
      sendError(res, 400, "filePath required");
      return;
    }
    const result = db.adminDeleteLock(filePath);
    if (result.ok) {
      hub.pushEvent("lock_released", { file_path: filePath });
      res.send(JSON.stringify({ deleted: true }));
    } else {
      sendError(res, 400, JSON.stringify(dbErrorToJson(result.error)));
    }
  });
};

/** Register POST /admin/delete-agent route. */
const registerDeleteAgentRoute = (
  app: Express,
  db: TooManyCooksDb,
  hub: AdminEventHub,
): void => {
  app.post("/admin/delete-agent", (req: Request, res: Response) => {
    const agentName = stringField(req.body as Record<string, unknown>, "agentName");
    if (agentName === undefined) {
      sendError(res, 400, "agentName required");
      return;
    }
    const result = db.adminDeleteAgent(agentName);
    if (result.ok) {
      hub.pushEvent("agent_deleted", { agent_name: agentName });
      res.send(JSON.stringify({ deleted: true }));
    } else {
      sendError(res, 400, JSON.stringify(dbErrorToJson(result.error)));
    }
  });
};

/** Register POST /admin/reset-key route. */
const registerResetKeyRoute = (
  app: Express,
  db: TooManyCooksDb,
): void => {
  app.post("/admin/reset-key", (req: Request, res: Response) => {
    const agentName = stringField(req.body as Record<string, unknown>, "agentName");
    if (agentName === undefined) {
      sendError(res, 400, "agentName required");
      return;
    }
    const result = db.adminResetKey(agentName);
    if (result.ok) {
      res.send(JSON.stringify(agentRegistrationToJson(result.value)));
    } else {
      sendError(res, 400, JSON.stringify(dbErrorToJson(result.error)));
    }
  });
};

/** Register POST /admin/send-message route. */
const registerSendMessageRoute = (
  app: Express,
  db: TooManyCooksDb,
  hub: AdminEventHub,
): void => {
  app.post("/admin/send-message", (req: Request, res: Response) => {
    const body = req.body as Record<string, unknown>;
    const fromAgent = stringField(body, "fromAgent");
    const toAgent = stringField(body, "toAgent");
    const content = stringField(body, "content");
    if (fromAgent === undefined || toAgent === undefined || content === undefined) {
      sendError(res, 400, "fromAgent, toAgent, content required");
      return;
    }
    const result = db.adminSendMessage(fromAgent, toAgent, content);
    if (result.ok) {
      hub.pushEvent("message_sent", {
        from_agent: fromAgent,
        to_agent: toAgent,
        message_id: result.value,
      });
      res.send(JSON.stringify({ sent: true, message_id: result.value }));
    } else {
      sendError(res, 400, JSON.stringify(dbErrorToJson(result.error)));
    }
  });
};

/** Register POST /admin/reset route. */
const registerResetRoute = (
  app: Express,
  db: TooManyCooksDb,
  hub: AdminEventHub,
): void => {
  app.post("/admin/reset", (_req: Request, res: Response) => {
    const result = db.adminReset();
    if (result.ok) {
      hub.pushEvent("state_reset", {});
      res.send(JSON.stringify({ reset: true }));
    } else {
      sendError(res, 500, JSON.stringify(dbErrorToJson(result.error)));
    }
  });
};

/** Register all admin routes on an Express app. */
export const registerAdminRoutes = (
  app: Express,
  db: TooManyCooksDb,
  hub: AdminEventHub,
): void => {
  app.use(express.json());
  registerStatusRoute(app, db);
  registerDeleteLockRoute(app, db, hub);
  registerDeleteAgentRoute(app, db, hub);
  registerResetKeyRoute(app, db);
  registerSendMessageRoute(app, db, hub);
  registerResetRoute(app, db, hub);
};
